### PR TITLE
SYCL Feature level 4 (parallel_for)

### PR DESCRIPTION
**Pull request by masterleinad**
Based on #3451. The most appropriate launch mechansim is still not quite clear so I am including the first one @nliber came up with (which has some severe restrictions on data types being trivially copyable to copy them to the device) and an indirect launch mechanism using shared memory.
I think that we can discuss further improvements to the lauch mechanism in a separate pull request.

### CODE REVIEW

**Comment by nliber** [comment]
@masterleinad this isn't the first one.  The one currently checked in here handles trivially copyable types by placement-newing them into USM shared memory.  It should functionally work although isn't the mechanism we'll be using.  I think our best course is to use this one for now unless problems show up, in which case update it to the latest.

**Review by nliber — APPROVED** [review]

**Comment by masterleinad** [comment]
Rebased.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp [hunk 1]
```diff
@@ -45,15 +45,16 @@
...
+#include <CL/sycl.hpp>
 
 namespace Kokkos {
 namespace Experimental {
 namespace Impl {
 
 class SYCLInternal {
  public:
-  using size_type = Kokkos::Experimental::SYCL::size_type;
+  using size_type = unsigned int;
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:57]: Why did you change this?

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:57]: This allowed me to resolve some circular dependency problems.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  template <class TagType>
+  typename std::enable_if<!std::is_same<TagType, void>::value>::type exec_range(
+      const Member i) const {
+    m_functor(TagType(), i);
+  }
+
+ public:
+  typedef FunctorType functor_type;
+
+  inline void operator()(cl::sycl::item<1> item) const {
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:84]: how does this work? The direct launch passes in an int?? Isn't that weird? Should direct_launch just pass on the sycl item?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    int id = item.get_linear_id();
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:85]: this needs to use the member/index_type from the policy.

**Review by crtrott — DISMISSED** [review]
Have a couple of questions and change requests.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:57]: I can defer to the Instance class to only define it in one place.

**Comment by masterleinad** [comment]
> can we please change this to a signed type. I ALWAYS regret that we made this unsigned on CUDA.

See https://github.com/kokkos/kokkos/pull/3484.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  q.wait();
+
+  auto end   = policy.end();
+  auto begin = policy.begin();
+  cl::sycl::range<1> range(end - begin);
+
+  q.submit([&](cl::sycl::handler& cgh) {
+    cgh.parallel_for(range, [=](cl::sycl::item<1> item) {
+      int id = item.get_linear_id();
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp:71]: We don't have any implicit arguments like `blockDim` and `gridDim` for `SYCL` so we have to due to the work mapping explicitly.
Currently, the `ParallelFor` call operator is not used and it's not quite clear to me if it needs to exist. On the other hand, it's also not quite clear to me if we can/want to generalize `KernelLaunch` in a way to support `parallel_for`, `parallel_reduce` and `parallel_scan`.
In #3480, everything is defined inline in `ParallelReduce` for example.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
+      *instance.m_indirectKernel;
+
+  // Allocate USM shared memory for the functor
+  kernelMem.resize(std::max(kernelMem.size(), sizeof(Functor)));
+
+  // Placement new a copy of functor into USM shared memory
+  //
+  // Store it in a unique_ptr to call its destructor on scope exit
+  std::unique_ptr<Functor, Kokkos::Impl::destruct_delete> kernelFunctorPtr(
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp:95]: At the moment, we fence in `sycl_direct_launch` and call the destructor at the end of this scope. We should reconsider this when refactoring the launch mechanism.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:84]: It's unused. Let me remove it for now to avoid confusion.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:85]: I'll just change to `auto`.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: the thing is if you want to reuse the kernel launch for TeamPolicy and MDRangePolicy you need to do something like what CUDA does, and consideirng that we have at least the direct vs indirect launch mechanism, and this is before we hit any of the optimizations we consider for CUDA like occupancy and what not I think we should go the same route here.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp]: no the expectation is that the user functor gets the member type from the policy.

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp [hunk 6]
```diff
@@ -45,15 +45,19 @@
 #ifndef KOKKOS_SYCL_INSTANCE_HPP_
 #define KOKKOS_SYCL_INSTANCE_HPP_
 
-#include <Kokkos_SYCL.hpp>
+#include <Kokkos_Macros.hpp>
+
+#if defined KOKKOS_ENABLE_SYCL
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.hpp:50]: Why does this need to be guarded at all?

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  q.wait();
+
+  auto end   = policy.end();
+  auto begin = policy.begin();
+  cl::sycl::range<1> range(end - begin);
+
+  q.submit([&](cl::sycl::handler& cgh) {
+    cgh.parallel_for(range, [=](cl::sycl::item<1> item) {
+      const typename Policy::index_type id = item.get_linear_id();
+      functor(id);
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp:72]: so I now get that this is actually ONLY for the parallel for dispatch. Since this is not reusable the way it is written why not make this part of the ParallelFor class? Or do we intend to modify this later to make it reusable?

File: core/src/SYCL/Kokkos_SYCL_Instance.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.hpp:50]: Because it's included by `Kokkos_SYCL_Space.hpp`. Let me just guard it there instead.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp (hunk 7 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: That's fine with me (and more in line with #3480). I pushed a corresponding commit. Maybe @nliber wants to comment?

**Comment by dalg24** [comment]
Please rewrite history and we will merge.  Make sure you make Nevin a co-author

**Comment by masterleinad** [comment]
> Please rewrite history and we will merge. Make sure you make Nevin a co-author

Here you go!

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: I agree that they should look the same. I guess `sycl_indirect_lauch` could possibly be shared but I think we can refactor when we at least also have `parallel_reduce`.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp (hunk 7 above)
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: I'd like to put all the launch stuff together, because if one of them needs to change, they probably all will.  It's only a very weak preference though.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp (hunk 3 above)
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp]: `sycl::handler::parallel_for`calls the function it is passed with a `sycl::item`.  We can't change what `item.get_linear_id()` returns (although it should be a `size_t`).  Should I just cast it before to the member type from the policy before passing it in?

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp]: This discussion is outdated. We are doing this already, see https://github.com/kokkos/kokkos/pull/3474/files#diff-3b75050857c4c9a6392a5b0dd07241be60a4fed7c8c9fa56ce2242a4324e551fR80. :slightly_smiling_face:
